crates/rethnet_eth/src/serde.rs

```diff
@@ -0,0 +1,165 @@
+/// helper utilities for use with serde's serialize_with and deserialize_with
+use std::{fmt::Write, ops::Deref};
+
+use crate::U256;
+
+/// Type that serializes a [`U256`] without leading zeroes.
+#[derive(Clone, Copy, serde::Serialize, serde::Deserialize)]
+pub struct U256WithoutLeadingZeroes(#[serde(serialize_with = "u256::serialize")] U256);
+
+impl Deref for U256WithoutLeadingZeroes {
+    type Target = U256;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<U256> for U256WithoutLeadingZeroes {
+    fn from(value: U256) -> Self {
+        Self(value)
+    }
+}
+
+/// Type that serializes a [`U256`] without leading zeroes.
+#[derive(Clone, Copy, serde::Serialize, serde::Deserialize)]
+pub struct U64WithoutLeadingZeroes(#[serde(serialize_with = "u64::serialize")] u64);
+
+impl Deref for U64WithoutLeadingZeroes {
+    type Target = u64;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+impl From<u64> for U64WithoutLeadingZeroes {
+    fn from(value: u64) -> Self {
+        Self(value)
+    }
+}
+
+/// for use with serde's serialize_with on a single value that should be serialized as a
+/// sequence
+pub fn single_to_sequence<S, T>(val: &T, s: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+    T: serde::Serialize,
+{
+    use serde::ser::SerializeSeq;
+    let mut seq = s.serialize_seq(Some(1))?;
+    seq.serialize_element(val)?;
+    seq.end()
+}
+
+/// for use with serde's deserialize_with on a sequence that should be deserialized as a
+/// single value
+pub fn sequence_to_single<'de, D, T>(deserializer: D) -> Result<T, D::Error>
+where
+    D: serde::Deserializer<'de>,
+    T: serde::Deserialize<'de> + Clone,
+{
+    let s: Vec<T> = serde::de::Deserialize::deserialize(deserializer)?;
+    Ok(s[0].clone())
+}
+
+/// Helper function for serializing the little-endian bytes of an unsigned integer into a hexadecimal string.
+fn serialize_uint_bytes_without_leading_zeroes<S, T>(le_bytes: T, s: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+    T: AsRef<[u8]>,
+{
+    let le_bytes = le_bytes.as_ref();
```

Comment:
Since we're using the generic trait `AsRef<[u8]>` we can't actually get big-endian bytes. We're converting those to big-endian bytes by reversing the iterator using `.rev()`